source/loader/layers/sanitizer/asan_interceptor.hpp

```diff
@@ -0,0 +1,138 @@
+        std::shared_lock<ur_shared_mutex> Guard(Mutex);
+        assert(AllocatedUSMMap.find(Address) != AllocatedUSMMap.end());
+        return AllocatedUSMMap[Address];
+    }
+
+    ur_shared_mutex Mutex;
+    // Note: nullptr is host device
+    std::unordered_map<ur_device_handle_t, std::unique_ptr<DeviceInfo>>
+        DeviceMap;
+    std::unordered_map<ur_queue_handle_t, std::unique_ptr<QueueInfo>> QueueMap;
+
+    /// key: USMAllocInfo.AllocBegin
+    /// value: USMAllocInfo
+    /// Use AllocBegin as key can help to detect underflow pointer
+    std::map<uptr, USMAllocInfo> AllocatedUSMMap;
+};
+
+class SanitizerInterceptor {
+  public:
+    SanitizerInterceptor() {}
+
+    ur_result_t allocateMemory(ur_context_handle_t Context,
+                               ur_device_handle_t Device,
+                               const ur_usm_desc_t *Properties,
+                               ur_usm_pool_handle_t Pool, size_t Size,
+                               void **ResultPtr, USMMemoryType Type);
+    ur_result_t releaseMemory(ur_context_handle_t Context, void *Ptr);
+
+    bool preLaunchKernel(ur_kernel_handle_t Kernel, ur_queue_handle_t Queue,
+                         ur_event_handle_t &Event);
+    void postLaunchKernel(ur_kernel_handle_t Kernel, ur_queue_handle_t Queue,
+                          ur_event_handle_t *Event, bool SetCallback = true);
+
+    ur_result_t addContext(ur_context_handle_t Context);
+    ur_result_t addDevice(ur_context_handle_t Context,
+                          ur_device_handle_t Device);
+    ur_result_t addQueue(ur_context_handle_t Context, ur_queue_handle_t Queue);
+
+  private:
+    ur_result_t updateShadowMemory(ur_queue_handle_t Queue);
+    ur_result_t enqueueAllocInfo(ur_context_handle_t Context,
+                                 ur_device_handle_t Device,
+                                 ur_queue_handle_t Queue,
+                                 USMAllocInfo &AlloccInfo,
+                                 ur_event_handle_t &LastEvent);
+
+    /// Initialize Global Variables & Kernel Name at first Launch
+    void prepareLaunch(ur_queue_handle_t Queue, ur_kernel_handle_t Kernel);
+
+    std::string getKernelName(ur_kernel_handle_t Kernel);
+    ur_result_t allocShadowMemory(ur_context_handle_t Context,
+                                  DeviceInfo &DeviceInfo);
+    ur_result_t enqueueMemSetShadow(ur_context_handle_t Context,
+                                    ur_device_handle_t Device,
+                                    ur_queue_handle_t Queue, uptr Addr,
+                                    uptr Size, u8 Value,
+                                    ur_event_handle_t DepEvent,
+                                    ur_event_handle_t *OutEvent);
+
+    ContextInfo &getContextInfo(ur_context_handle_t Context) {
+        std::shared_lock<ur_shared_mutex> Guard(m_ContextMapMutex);
+        assert(m_ContextMap.find(Context) != m_ContextMap.end());
+        return *m_ContextMap[Context].get();
```

Comment:
You are dropping the lock and returning a reference to an underlying value of a unique_ptr. While I think this is safe with the current implementation, because these things are only ever removed in the destructor, it might be worth considering using a shared_ptr here.

Reply:
Makes sense. I decided to use shared_ptr, thanks